File: 8/1/dominator.R
# Find an index of an array such that its value occurs at more than half of indices in the array.
dominator <- function(vec_a){
  if(length(vec_a)==0){
    return(-1)
  }
  # use the table function to count the occurrences of each element
  occur <- table(vec_a)
  # use the max function to obtain the value that occurres the most
  value <- as.numeric(names(occur[occur==max(occur)]))
  # list all the indices that correspond to value
  indices <- which(vec_a == value)
  # if length of indices is less than half of length of vec_a, return -1
  if(length(indices) * 2 <= length(vec_a)){
    return(-1)
  }
  # return any of the indices (adjusted to zero-index results)
  return(sample(indices, 1) - 1)
}
# debugging
# VEC_A <- c(3, 4, 3, 2, 3, -1, 3, 3)
# print(dominator(VEC_A) %in% c(0, 2, 4, 6, 7))
# expect degit that corresponds as zero-index to value 3
# Bash Testing:
VEC_A <- commandArgs()[6] # store arguments passed from command line
VEC_A <- gsub("\\]|\\[", "", VEC_A) # remove brackets
VEC_A <- unlist(strsplit(VEC_A, ", *")) # conver to vector
print(dominator(VEC_A))

File: 8/1/dominator.js
// Find an index of an array such that its value occurs at more than half of indices in the array.
function dominator(arr_a){
    // create objects to track recurrences of each element in array
    var counts = {};
    // iterate through array
    for (var ndx_i = 0; ndx_i < arr_a.length; ndx_i++) {
        // test whether counts object lists element ? if so, add one : if not, set at one
        counts[arr_a[ndx_i]] = counts[arr_a[ndx_i]] ? counts[arr_a[ndx_i]] + 1 : 1;
        // test whether more than half the array matches the count of elements
        if(arr_a.length < counts[arr_a[ndx_i]] * 2){
            // if so, return ndx_i
            return(ndx_i);
        }
    }
    return(-1);
}
// Debugging:
ARGS_A = [3,4,3,2,3,-1,3,3];
dominator(ARGS_A);
// Bash testing:
var ARGS_A = process.argv[2]; // read arguments from command line
ARGS_A = ARGS_A.replace(/\[|\]/g,''); // remove brackets
ARGS_A = ARGS_A.split(/, */); // split by commas
ARGS_A = ARGS_A.map(Number);
// eslint-disable-next-line no-console
console.log(dominator(ARGS_A));
// Codility testing:
// https://app.codility.com/demo/results/trainingP29RER-XEA/
// Correctness: 100%
// Performance: 100%
// Difficulty: Painless
